### PR TITLE
Notifications api

### PR DESCRIPTION
**Pull request by robsdedude**
Depends on
 - https://github.com/neo4j-drivers/testkit/pull/533

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,10 @@
 # Neo4j Driver Change Log (breaking/major changes only)
 
+## Version 5.x
```
Comment (robsdedude) [inline CHANGELOG.md]: :warning: Before merging: don't forget to finalize!

File: docs/source/api.rst [hunk 2]
```diff
@@ -370,6 +371,25 @@ Specify the client agent name.
...
+
+This can be a single or an iterable of :class:`.NotificationFilter` object(s) or
+equivalent string(s) (see :class:`.NotificationFilter`).
+Alternatively, :const:`None` can be used to indicate
+:attr:`NotificationFilter.DEFAULT`.
+
+:Type: :const:`None`, :class:`.NotificationFilter`, :class:`str`, or iterable of either
+:Default: :const:`None`
+
+.. versionadded:: 5.2
```
Comment (robsdedude) [inline docs/source/api.rst]: :warning: Before merging: don't forget to finalize!

File: docs/source/api.rst [hunk 3]
```diff
@@ -685,6 +706,24 @@ See :class:`.BookmarkManager` for more information.
...
+
+This can be a single or an iterable of :class:`.NotificationFilter` object(s) or
+equivalent string(s) (see :class:`.NotificationFilter`).
+Alternatively, :const:`None` can be used fall back to the filters configured on
+the driver (see :ref:`driver-notification-filters-ref`).
+
+:Type: :const:`None`, :class:`.NotificationFilter`, :class:`str`, or iterable of either
+:Default: :const:`None`
+
+.. versionadded:: 5.2
```
Comment (robsdedude) [inline docs/source/api.rst]: :warning: Before merging: don't forget to finalize!

File: neo4j/_api.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        True
+        >>> INFORMATION_PERFORMANCE == "INFORMATION.PERFORMANCE"
+        True
+
+    :attr:`NONE` instructs the server to not send any notifications.
+    :attr:`DEFAULT` leaves it up to the server which notifications to send.
+
+    The other choices are a combination of a severity and a category.
+
+    .. versionadded:: 5.2
```
Comment (robsdedude) [inline neo4j/_api.py]: :warning: Before merging: don't forget to finalize!

File: neo4j/_api.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            "INFORMATION.QUERY",
+            "INFORMATION.PERFORMANCE",
+        ],
+    ]
+
+
+class NotificationSeverity(Enum):
+    """Server-side notification severity.
+
+    .. versionadded:: 5.2
```
Comment (robsdedude) [inline neo4j/_api.py]: :warning: Before merging: don't forget to finalize!

File: neo4j/_api.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    WARNING = "WARNING"
+    INFORMATION = "INFORMATION"
+    #: Used when server provides a Severity which the driver is unaware of.
+    UNKNOWN = "UNKNOWN"
+
+
+class NotificationCategory(Enum):
+    """Server-side notification category.
+
+    .. versionadded:: 5.2
```
Comment (robsdedude) [inline neo4j/_api.py]: :warning: Before merging: don't forget to finalize!

File: tests/unit/common/work/test_summary.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import typing as t
+
+
```
Comment (robsdedude) [inline tests/unit/common/work/test_summary.py:21]:
```suggestion
```

File: bin/make-unasync [hunk 8]
```diff
@@ -223,6 +223,7 @@ def apply_unasync(files):
         "assert_awaited_once": "assert_called_once",
         "assert_awaited_once_with": "assert_called_once_with",
         "await_count": "call_count",
+        "AsyncMock": "MagicMock",
```
Comment (bigmontz) [inline bin/make-unasync:226]: Where did MagicMock come from?

File: src/neo4j/_work/summary.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+            kwargs["raw_severity_level"] = severity
+            kwargs["severity_level"] = _SEVERITY_LOOKUP.get(
+                severity, NotificationSeverity.UNKNOWN
+            )
+        category = metadata.get("category")
+        if isinstance(category, str):
+            kwargs["raw_category"] = category
+            kwargs["category"] = _CATEGORY_LOOKUP.get(
+                category, NotificationCategory.UNKNOWN
+            )
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py:275]: You are overriding category with a value which come direct from the wire. This value should be UNKNOWN by default like `severity_level`.

**Review by bigmontz — CHANGES_REQUESTED** [review]
The behaviour of the property `category` of the ResultSummary is wrong, it should be `UNKNOWN` when the category is not present in the metadata.

File: src/neo4j/_work/summary.py (hunk 9 above)
Comment (bigmontz) [reply src/neo4j/_work/summary.py:275]: The code was missread.

**Review by bigmontz — APPROVED** [review]
🔢

File: bin/make-unasync (hunk 8 above)
Comment (robsdedude) [reply bin/make-unasync:226]: the mocking library has `Mock` and `MagicMock` which behave the same except that the latter also mock magic methods (`__xyz__` shaped methods). `AsyncMock` does always mock magic methods as well (I suppose it has to, to be able to intercept calls to `__await__`, but that is speculation). Either way, by default the unasync script would turn `AsyncMock` into `Mock` which brakes tests that rely on magic methods being mocked.

**Pull request merged**